components/zcash_note_encryption/src/lib.rs

```diff
@@ -1,8 +1,471 @@
+            esk,
+            note,
+            to,
+            memo,
+            ovk,
+        }
+    }
+
+    /// Exposes the ephemeral secret key being used to encrypt this note.
+    pub fn esk(&self) -> &D::EphemeralSecretKey {
+        &self.esk
+    }
+
+    /// Exposes the encoding of the ephemeral public key being used to encrypt this note.
+    pub fn epk(&self) -> &D::EphemeralPublicKey {
+        &self.epk
+    }
+
+    /// Generates `encCiphertext` for this note.
+    pub fn encrypt_note_plaintext(&self) -> [u8; ENC_CIPHERTEXT_SIZE] {
+        let pk_d = D::get_pk_d(&self.note);
+        let shared_secret = D::ka_agree_enc(&self.esk, &pk_d);
+        let key = D::kdf(shared_secret, &self.epk);
+        let input = D::note_plaintext_bytes(&self.note, &self.to, &self.memo);
+
+        let mut output = [0u8; ENC_CIPHERTEXT_SIZE];
+        assert_eq!(
+            ChachaPolyIetf::aead_cipher()
+                .seal_to(&mut output, &input.0, &[], key.as_ref(), &[0u8; 12])
+                .unwrap(),
+            ENC_CIPHERTEXT_SIZE
+        );
+
+        output
+    }
+
+    /// Generates `outCiphertext` for this note.
+    pub fn encrypt_outgoing_plaintext<R: RngCore>(
+        &self,
+        cv: &D::ValueCommitment,
+        cm: &D::NoteCommitment,
+        rng: &mut R,
+    ) -> [u8; OUT_CIPHERTEXT_SIZE] {
+        let (ock, input) = if let Some(ovk) = &self.ovk {
+            let ock = D::get_ock(ovk, &cv, &cm, &self.epk);
+            let input = D::outgoing_plaintext_bytes(&self.note, &self.esk);
+
+            (ock, input)
+        } else {
+            // ovk = ⊥
+            let mut ock = OutgoingCipherKey([0; 32]);
+            let mut input = [0u8; OUT_PLAINTEXT_SIZE];
+
+            rng.fill_bytes(&mut ock.0);
+            rng.fill_bytes(&mut input);
+
+            (ock, OutPlaintextBytes(input))
+        };
+
+        let mut output = [0u8; OUT_CIPHERTEXT_SIZE];
+        assert_eq!(
+            ChachaPolyIetf::aead_cipher()
+                .seal_to(&mut output, &input.0, &[], ock.as_ref(), &[0u8; 12])
+                .unwrap(),
+            OUT_CIPHERTEXT_SIZE
+        );
+
+        output
+    }
+}
+
+/// Trial decryption of the full note plaintext by the recipient.
+///
+/// Attempts to decrypt and validate the given `enc_ciphertext` using the given `ivk`.
+/// If successful, the corresponding Sapling note and memo are returned, along with the
+/// `PaymentAddress` to which the note was sent.
+///
+/// Implements section 4.17.2 of the Zcash Protocol Specification.
+pub fn try_note_decryption<D: Domain>(
+    domain: &D,
+    //output: &ShieldedOutput<D>,
+    ivk: &D::IncomingViewingKey,
+    epk: &D::EphemeralPublicKey,
+    cmstar: &D::ExtractedCommitment,
+    enc_ciphertext: &[u8],
+) -> Option<(D::Note, D::Recipient, D::Memo)> {
+    assert_eq!(enc_ciphertext.len(), ENC_CIPHERTEXT_SIZE);
+
+    let shared_secret = D::ka_agree_dec(ivk, epk);
+    let key = D::kdf(shared_secret, epk);
```

Comment:
Here, KA.Agree takes the decoded epk while the KDF takes `ephemeralKey`.

Reply:
You could also argue that `check_note_validity` would reject non-canonical `ephemeralKey` since it compares byte sequences. It's up to you whether to rely on that or follow the spec more literally.